core/src/repair/malicious_repair_handler.rs

```diff
@@ -0,0 +1,128 @@
+use {
+    super::{
+        repair_handler::RepairHandler, repair_response::repair_response_packet_from_bytes,
+        standard_repair_handler::StandardRepairHandler,
+    },
+    solana_clock::Slot,
+    solana_ledger::{
+        blockstore::Blockstore,
+        shred::{Nonce, SIZE_OF_DATA_SHRED_HEADERS},
+    },
+    solana_perf::packet::{Packet, PacketBatch, PacketBatchRecycler, PinnedPacketBatch},
+    std::{net::SocketAddr, sync::Arc},
+};
+
+#[derive(Copy, Clone, Debug, Default)]
+pub struct MaliciousRepairConfig {
+    bad_shred_slot_frequency: Option<Slot>,
+}
+
+pub struct MaliciousRepairHandler {
+    blockstore: Arc<Blockstore>,
+    standard_handler: StandardRepairHandler,
+    config: MaliciousRepairConfig,
+}
+
+impl MaliciousRepairHandler {
+    const BAD_DATA_INDEX: usize = SIZE_OF_DATA_SHRED_HEADERS + 5;
+
+    pub fn new(blockstore: Arc<Blockstore>, config: MaliciousRepairConfig) -> Self {
+        let standard_handler = StandardRepairHandler::new(blockstore.clone());
+        Self {
+            blockstore,
+            standard_handler,
+            config,
+        }
+    }
+
+    fn repair_response_packet(
+        &self,
+        slot: Slot,
+        shred_index: u64,
+        dest: &SocketAddr,
+        nonce: Nonce,
+    ) -> Option<Packet> {
+        let mut shred = self
+            .blockstore
+            .get_data_shred(slot, shred_index)
+            .expect("Blockstore could not get data shred")?;
+        if self
+            .config
+            .bad_shred_slot_frequency
+            .is_some_and(|freq| slot % freq == 0)
+        {
+            // Change some random piece of data
+            shred[Self::BAD_DATA_INDEX] = shred[Self::BAD_DATA_INDEX].wrapping_add(1);
+        }
+        repair_response_packet_from_bytes(shred, dest, nonce)
+    }
+}
+
+impl RepairHandler for MaliciousRepairHandler {
+    fn run_window_request(
+        &self,
+        recycler: &PacketBatchRecycler,
+        from_addr: &SocketAddr,
+        slot: Slot,
+        shred_index: u64,
+        nonce: Nonce,
+    ) -> Option<PacketBatch> {
+        let packet = self.repair_response_packet(slot, shred_index, from_addr, nonce)?;
```

Comment:
Maybe we could define `repair_response_packet` as a trait method and then the `run_window_request` can be a default trait implementation that calls `repair_response_packet`

```
trait RepairHandler {
   fn repair_response_packet()

    fn run_window_request() {
        let packet = self.repair_response_packet;
        ...
    }
}
```

Reply:
yeah good call I rolled a bunch of stuff into the trait to keep the implementations simple 976118511aad3f63c23b000055b82a56b3d17933